### PR TITLE
Add CT verion check against talker requirements

### PR DESCRIPTION
**Pull request by mizaki**
I picked `>=1.6.0a7` as the default because that is when the lists were changed to sets etc.

### CODE REVIEW

**Comment by mizaki** [comment]
Should `ct_req_spec` be an empty string in class `ComicTalker`?

Should the GUI notify if a talker fails to load?

File: comictalker/comictalker.py [hunk 1]
```diff
@@ -107,6 +107,7 @@ class ComicTalker:
 
     name: str = "Example"
     id: str = "example"
+    ct_req_spec: str = ">=1.6.0a7"  # The ComicTagger version required by the talker using PyPA version specifiers
```
Comment (lordwelch) [inline comictalker/comictalker.py]: ct should be spelled out

File: comictalker/__init__.py [hunk 2]
```diff
@@ -30,7 +32,15 @@ def get_talkers(version: str, cache: pathlib.Path) -> dict[str, ComicTalker]:
...
-            talkers[talker.name] = obj
+            try:
+                if version in SpecifierSet(obj.ct_req_spec, prereleases=True):
+                    talkers[talker.name] = obj
+                else:
+                    logger.error(
+                        f"CT required version not met for talker: {talker.name} with specifier: {obj.ct_req_spec}"
+                    )
+            except InvalidSpecifier:
+                logger.error(f"Invalid specifier for talker: {talker.name} - specifier: {obj.ct_req_spec}")
```
Comment (lordwelch) [inline comictalker/__init__.py]: How about we just make it minimum version instead of using the full specification

Comment (mizaki) [reply comictalker/__init__.py]: My thinking was in case there are breaking changes a max version could be set.

Comment (lordwelch) [reply comictalker/__init__.py]: My rationale is that if there is a breaking change they will have to make a code change to exclude it, and instead they should make the needed code changes

File: comictalker/__init__.py [hunk 3]
```diff
@@ -30,7 +32,15 @@ def get_talkers(version: str, cache: pathlib.Path) -> dict[str, ComicTalker]:
...
-            talkers[talker.name] = obj
+            try:
+                if parse(version) >= parse(obj.comictagger_min_ver):
+                    talkers[talker.name] = obj
+                else:
+                    logger.error(
+                        f"CT minimum required version not met for talker: {talker.name} with version: {obj.comictagger_min_ver}"
+                    )
+            except InvalidVersion:
+                logger.error(f"Invalid version number for talker: {talker.name} - version: {obj.comictagger_min_ver}")
```
Comment (lordwelch) [inline comictalker/__init__.py]: I think this should be a warning instead and we should still try to use the plugin

File: comictalker/__init__.py [hunk 4]
```diff
@@ -30,7 +32,15 @@ def get_talkers(version: str, cache: pathlib.Path) -> dict[str, ComicTalker]:
             if obj.id != talker.name:
                 logger.error("Talker ID must be the same as the entry point name")
                 continue
-            talkers[talker.name] = obj
+            try:
+                if parse(version) >= parse(obj.comictagger_min_ver):
+                    talkers[talker.name] = obj
+                else:
+                    logger.error(
+                        f"CT minimum required version not met for talker: {talker.name} with version: {obj.comictagger_min_ver}"
```
Comment (lordwelch) [inline comictalker/__init__.py]: CT should be spelled out. To me this reads that the version is the talker version and not ComicTagger's version

**Review by lordwelch — COMMENTED** [review]
We should list the plugins that are disabled and why eg execution failed, wrong version but doesn't need to be implemented in this PR

**Comment by lordwelch** [comment]
I don't like pulling in the packaging dependency when we just need to parse a basic version
Something like this should fit our needs
```python
import pytest
from typing import Tuple

def parse_version(version_str: str) -> Tuple[int, int, int]:
    # Split the version string into segments and convert them to integers
    version_parts = version_str.split('.')
    parts = [int(seg) if seg.isdigit() else 0 for seg in version_parts[:3]]
    
    # Ensure exactly three elements in the resulting list
    parts.extend([0] * (3 - len(parts)))

    # Return a tuple with the first three elements from the parsed segments
    return (parts[0], parts[1], parts[2])

@pytest.mark.parametrize(
    "version_str, expected",
    [
        ("1.2.3", (1, 2, 3)),       # Test case with a standard version string
        ("0.0.1", (0, 0, 1)),       # Test case with segments containing only zeros except for the last segment
        ("10.5", (10, 5, 0)),       # Test case with only two segments
        ("2", (2, 0, 0)),           # Test case with only one segment
        ("3.7.9.2", (3, 7, 9)),     # Test case with more than three segments
        ("", (0, 0, 0))             # Test case for an empty string
    ],
)
def test_parse_version(version_str, expected):
    # Check if the output of parse_version matches the expected value for each test case
    assert parse_version(version_str) == expected
```

Tuples can then be compared with <,> easily

**Comment by mizaki** [comment]
> I don't like pulling in the packaging dependency when we just need to parse a basic version Something like this should fit our needs

Given the current importance of the 'a<number>' parsing that out to a 4 part number seems sensible?

**Comment by lordwelch** [comment]
You're right, I forgot about the alpha releases, those need to sort before the normal release. Then we'll keep the packaging dependency.

File: comictalker/__init__.py [hunk 5]
```diff
@@ -30,7 +32,20 @@ def get_talkers(version: str, cache: pathlib.Path) -> dict[str, ComicTalker]:
             if obj.id != talker.name:
                 logger.error("Talker ID must be the same as the entry point name")
                 continue
-            talkers[talker.name] = obj
+            try:
+                if parse(version) >= parse(obj.comictagger_min_ver):
```
Comment (lordwelch) [inline comictalker/__init__.py]: Move this `parse(version)` out of the loop, we don't need to parse this again for every plugin

**Review by lordwelch — CHANGES_REQUESTED** [review]

**Review by lordwelch — APPROVED** [review]

**Pull request merged**